apps/prairielearn/src/lib/submissions-upload.ts

```diff
@@ -0,0 +1,228 @@
+import csvtojson from 'csvtojson';
+import { parseISO } from 'date-fns';
+import memoize from 'p-memoize';
+import * as streamifier from 'streamifier';
+import { z } from 'zod';
+
+import * as sqldb from '@prairielearn/postgres';
+
+import { selectAssessmentInfoForJob } from '../models/assessment.js';
+import { selectQuestionByQid } from '../models/question.js';
+import { selectOrInsertUserByUid } from '../models/user.js';
+
+import { deleteAllAssessmentInstancesForAssessment } from './assessment.js';
+import { AssessmentQuestionSchema, IdSchema } from './db-types.js';
+import { createServerJob } from './server-jobs.js';
+
+const sql = sqldb.loadSqlEquiv(import.meta.url);
+
+const ZodStringToJson = z.preprocess((val) => {
+  if (val === '' || val == null) return {};
+  return JSON.parse(String(val));
+}, z.record(z.any()).nullable());
+
+const SubmissionCsvRowSchema = z.object({
+  UID: z.string(),
+  // We only use this if someone tries to upload a CSV for a group assessment,
+  // in which case we'll throw an error.
+  'Group name': z.string().optional(),
+  'Assessment instance': z.coerce.number().int(),
+  Question: z.string(),
+  Variant: z.coerce.number().int(),
+  Seed: z.string(),
+  Params: ZodStringToJson,
+  'True answer': ZodStringToJson,
+  Options: ZodStringToJson,
+  'Submission date': z
+    .string()
+    // We pipe the value through `parseISO` from `date-fns` because our CSV
+    // export uses timezone offsets like `-05` which, while valid per the
+    // ISO 8601 standard, aren't parseable by JavaScript's `Date` constructor
+    // or Zod's own date coercion.
+    .transform((val) => parseISO(val))
+    .pipe(z.date()),
+  'Submitted answer': ZodStringToJson,
+});
+
+/**
+ * A utility function to help with inserting only a single row per combination of keys.
+ */
+function makeDedupedInserter<T>() {
```

Comment:
Can you clarify what this function does that is different than `memoize`? Or more specifically, why you don't use `memoize` in the implementation of this function?